pandasql/cost_model.py

```diff
@@ -1,57 +1,104 @@
 from pandasql.utils import _get_dependency_graph
 import pandasql as ps
 
+
 class CostModel(object):
-    def __init__(self, max_pandas_len=10, max_pandas_input=500_000_000 ): #500 MB
-        self.transfer_cost = 0 # GB/s
+    def __init__(self, max_pandas_len=10,
+                 max_pandas_input=500_000_000):  # 500 MB
+        self.transfer_rate = 0  # GB/s
         self.required_pandas_ops = []
         self.required_sql_ops = []
-        self.system_memory_size = 0
+        self.system_memory_size = 0  # TODO: automate finding this
         self.max_pandas_len = max_pandas_len
         self.max_pandas_input = max_pandas_input
         self.compute_path = []
 
     def should_offload(self, df):
-
-        # possible_executions = []
         graph = _get_dependency_graph(df)
-        query_height = len(graph)
-        #TODO: Only sum input_sizes for nodes that we compute from (don't double count)
-        estimated_input = sum([g.memory_usage for g in graph])
-
-        limits = [g.n for g in graph if isinstance(g, ps.core.Limit)]
-        filters = [g.criterion for g in graph if isinstance(g, ps.core.Selection)]
-        aggregations = [g.agg for g in graph if isinstance(g, ps.core.Aggregator)]
-        joins = [g.left_keys for g in graph if isinstance(g, ps.core.Join)]
-
-        # print(num_limits, num_filters, num_aggregations, num_joins)
-        print([type(g) for g in graph])
-
-        print(estimated_input, self.max_pandas_input)
-
-        # if query_height > self.max_pandas_len:
-        #     print('offloading from query height')
-        #     return True
-
-        if estimated_input > self.max_pandas_input:
-            print('offloading from input size')
-            return True
-
-        if len(aggregations) + len(joins) > 2:
-            print('offloading from agg/join')
-            return True
-
-        if len(limits) > 0 and min(limits) < 10000:
-            print('offloading from limit')
-            return True
-
-
-        else:
-            return False
-        # execution_estimates = []
-        # for pe in possible_executions:
-            # cost = 0 # some formula
-            # execution time + possible data transfer time
-            # we in
-            # execution_estimates.append(pe, cost)
-        # best_plan = min(execution_estimates, key = lambda t: t[1])
+
+        for rule in OFFLOADING_RULES:
```

Comment:
Not for now, but perhaps in the future we might want to structure this so that we can have a threshold. For example, if we fail more than N rules then we do X.

Reply:
Good idea. Will leave for future.